openstack/blockstorage/storagepools/results.go

```diff
@@ -0,0 +1,122 @@
+package storagepools
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+
+	"github.com/gophercloud/gophercloud"
+)
+
+type ListResult struct {
+	gophercloud.Result
+}
+
+// Minimum set of driver capabilities only
+// https://github.com/openstack/cinder/blob/master/cinder/interface/volume_driver.py#L56
+type Capabilities struct {
+	// Required Fields
+	DriverVersion     string  `json:"driver_version"`
+	FreeCapacityGB    float64 `json:"-"`
+	StorageProtocol   string  `json:"storage_protocol"`
+	TotalCapacityGB   float64 `json:"-"`
+	VendorName        string  `json:"vendor_name"`
+	VolumeBackendName string  `json:"volume_backend_name"`
+	// Optional Fields
+	ReservedPercentage       int64   `json:"reserved_percentage"`
+	LocationInfo             string  `json:"location_info"`
+	QoSSupprt                bool    `json:"QoS_support"`
+	ProvisonedCapacityGB     float64 `json:"provisoned_capacity_gb"`
+	MaxOverSubscriptionRatio float64 `json:"max_over_subscription_ratio"`
+	ThinProvisioningSupport  bool    `json:"thin_provisioning_support"`
+	ThickProvisioningSupport bool    `json:"thick_provisioning_support"`
+	TotalVolumes             int64   `json:"total_volumes"`
+	FilterFunction           string  `json:"filter_function"`
+	GoodnessFuction          string  `json:"goodness_function"`
+	Mutliattach              bool    `json:"multiattach"`
+	SparseCopyVolume         bool    `json:"sparse_copy_volume"`
+}
+
+type StoragePool struct {
+	Name         string `json:"name"`
+	Capabilities `json:"capabilities"`
+}
+
+func (r *StoragePool) UnmarshalJSON(b []byte) error {
+	// Unmarshal the generic stuff
+	type tmp StoragePool
+	var s *struct {
+		tmp
+	}
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+	*r = StoragePool(s.tmp)
+
+	// Unmarshal the more complex things
+	var q *struct {
+		Capabilities struct {
+			FreeCapacityGB  interface{} `json:"free_capacity_gb"`
+			TotalCapacityGB interface{} `json:"total_capacity_gb"`
+		} `json:"capabilities"`
+	}
+	err = json.Unmarshal(b, &q)
+	if err != nil {
+		return err
+	}
+
+	// Should be a numeric, "unknown", "infinite"
+	if q.Capabilities.FreeCapacityGB != nil {
+		switch t := q.Capabilities.FreeCapacityGB.(type) {
+		case float64:
+			r.Capabilities.FreeCapacityGB = q.Capabilities.FreeCapacityGB.(float64)
+		case int:
+			r.Capabilities.FreeCapacityGB = q.Capabilities.FreeCapacityGB.(float64)
```

Comment:
That will panic. Type-assert to `int`, then cast to `float64`. Better would be writing a unit test for it.